File: crates/tethys-api/src/skills.rs
//! `skills.*` namespace (`architecture.md` §12.1).

use tethys_schema::sync::{
    Scope, SkillImportSource, SkillInfo, SkillUpdateApplied, SkillUpdateCheck, SkillUpdatePlan,
};

use crate::ApiError;

/// Skill library methods.
pub trait SkillsApi: Send + Sync {
    fn skills_list(
        &self,
        _project_root: String,
    ) -> impl std::future::Future<Output = Result<Vec<SkillInfo>, ApiError>> + Send {
        async { Err(ApiError::Unimplemented("skills.list")) }
    }

    fn skills_import(
        &self,
        _project_root: String,
        _scope: Scope,
        _source: SkillImportSource,
    ) -> impl std::future::Future<Output = Result<SkillInfo, ApiError>> + Send {
        async { Err(ApiError::Unimplemented("skills.import")) }
    }

    fn skills_update_check(
        &self,
        _project_root: String,
        _scope: Scope,
        _name: String,
    ) -> impl std::future::Future<Output = Result<SkillUpdateCheck, ApiError>> + Send {
        async { Err(ApiError::Unimplemented("skills.update_check")) }
    }

    fn skills_update_plan(
        &self,
        _project_root: String,
        _scope: Scope,
        _name: String,
    ) -> impl std::future::Future<Output = Result<SkillUpdatePlan, ApiError>> + Send {
        async { Err(ApiError::Unimplemented("skills.update_plan")) }
    }

    fn skills_update_apply(
        &self,
        _project_root: String,
        _scope: Scope,
        _name: String,
    ) -> impl std::future::Future<Output = Result<SkillUpdateApplied, ApiError>> + Send {
        async { Err(ApiError::Unimplemented("skills.update_apply")) }
    }

    fn skills_trust(
        &self,
        _project_root: String,
        _scope: Scope,
        _name: String,
    ) -> impl std::future::Future<Output = Result<SkillInfo, ApiError>> + Send {
        async { Err(ApiError::Unimplemented("skills.trust")) }
    }

    fn skills_enable(
        &self,
        _project_root: String,
        _scope: Scope,
        _name: String,
        _enabled: bool,
    ) -> impl std::future::Future<Output = Result<SkillInfo, ApiError>> + Send {
        async { Err(ApiError::Unimplemented("skills.enable")) }
    }
}

File: packages/client/src/namespaces/skills.ts
//! `skills.*` client namespace. Wave 2 owner: D (M1.11).

import type {
  Scope,
  SkillImportSource,
  SkillInfo,
  SkillUpdateApplied,
  SkillUpdateCheck,
  SkillUpdatePlan,
} from "@tethys/bindings";

import type { Call } from "../transport";

export function skillsNamespace(call: Call) {
  return {
    list: (workspaceId: string) =>
      call<SkillInfo[]>("skills_list", { workspaceId }),
    import: (workspaceId: string, scope: Scope, source: SkillImportSource) =>
      call<SkillInfo>("skills_import", { workspaceId, scope, source }),
    updateCheck: (workspaceId: string, scope: Scope, name: string) =>
      call<SkillUpdateCheck>("skills_update_check", {
        workspaceId,
        scope,
        name,
      }),
    update_check: (workspaceId: string, scope: Scope, name: string) =>
      call<SkillUpdateCheck>("skills_update_check", {
        workspaceId,
        scope,
        name,
      }),
    updatePlan: (workspaceId: string, scope: Scope, name: string) =>
      call<SkillUpdatePlan>("skills_update_plan", {
        workspaceId,
        scope,
        name,
      }),
    update_plan: (workspaceId: string, scope: Scope, name: string) =>
      call<SkillUpdatePlan>("skills_update_plan", {
        workspaceId,
        scope,
        name,
      }),
    updateApply: (workspaceId: string, scope: Scope, name: string) =>
      call<SkillUpdateApplied>("skills_update_apply", {
        workspaceId,
        scope,
        name,
      }),
    update_apply: (workspaceId: string, scope: Scope, name: string) =>
      call<SkillUpdateApplied>("skills_update_apply", {
        workspaceId,
        scope,
        name,
      }),
    trust: (workspaceId: string, scope: Scope, name: string) =>
      call<SkillInfo>("skills_trust", { workspaceId, scope, name }),
    enable: (
      workspaceId: string,
      scope: Scope,
      name: string,
      enabled: boolean,
    ) =>
      call<SkillInfo>("skills_enable", {
        workspaceId,
        scope,
        name,
        enabled,
      }),
  };
}
